pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -39,284 +108,413 @@ func Add(mgr manager.Manager) error {
-	if err = drainPods(r, node, stop); err != nil {
-		return r.reconcileAndError(instance, err)
+func (r *ReconcileNodeMaintenance) setStatusAnnotation(node *corev1.Node, statusInfo *ReconcileNodeMaintenanceStatusInfo) (*corev1.Node, error) {
+	if data, err := json.Marshal(statusInfo); err != nil {
+		return nil, fmt.Errorf("invalid value of %s %v", NodeMaintenanceStatusAnnotation, err)
+	} else {
+
+		newNode := node.DeepCopy()
+		newNode.Annotations[NodeMaintenanceStatusAnnotation] = string(data)
+
+		err = r.patchNodes(node, newNode)
+		if err != nil {
+			log.Error("Can't set status annotation of node %v", err)
+		}
+
+		return newNode, err
 	}
+}
 
-	instance.Status.Phase = kubevirtv1alpha1.MaintenanceSucceeded
-	instance.Status.PendingPods = nil
-	err = r.client.Status().Update(context.TODO(), instance)
-	if err != nil {
-		reqLogger.Errorf("Failed to update NodeMaintenance with \"Succeeded\" status. Error: %v", err)
-		return r.reconcileAndError(instance, err)
+func (r *ReconcileNodeMaintenance) doesLeaseExist(client client.Client, nodeName string) (int, error, *coordv1beta1.Lease) {
+	lease := &coordv1beta1.Lease{}
+
+	nName := apitypes.NamespacedName{Namespace: corev1.NamespaceNodeLease, Name: nodeName}
+
+	if err := client.Get(context.TODO(), nName, lease); err != nil {
+		if errors.IsNotFound(err) {
+			return LeaseStatusNotFound, nil, nil
+		}
+		return LeaseStatusFail, err, nil
 	}
 
-	return reconcile.Result{}, nil
-}
+	// check if the existing lease is currently active
+	leaseDeadline := (*lease.Spec.AcquireTime).Time
+	leaseDeadline.Add(time.Duration(*lease.Spec.LeaseDurationSeconds) * time.Second)
 
-func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
-	node, err := r.fetchNode(nodeName)
-	if err != nil {
-		return err
+	if leaseDeadline.After(time.Now()) {
+		// lease active
+
+		if nodeName == *lease.Spec.HolderIdentity {
+			leaseDeadline.Add(time.Duration(-int64(LeaseGracePeriod) * int64(time.Second)))
+			if leaseDeadline.Before(time.Now()) {
+				return LeaseStatusActiveOwnedByMeAboutToExpire, nil, lease
+			}
+
+			return LeaseStatusActiveOwnedByMe, nil, lease
+		} else {
+			return LeaseStatusActiveOwnedByDifferentNode, nil, lease
+		}
 	}
 
-	// Uncordon the node
-	err = AddOrRemoveTaint(r.drainer.Client, node, false)
-	if err != nil {
+	// existing lease expired.
+	return LeaseStatusExpired, nil, nil
+}
+
+func (r *ReconcileNodeMaintenance) createLease(nodeName string, requestedMaintenenceWindowSeconds int32) error {
+
+	if requestedMaintenenceWindowSeconds < MinMaintenenceWindowSeconds {
+		err := fmt.Errorf("node %s maintenance window size too small", nodeName)
+		r.reqLogger.Errorf("%v", err)
 		return err
 	}
+	tmpDurationInSeconds := requestedMaintenenceWindowSeconds
+	lease := &coordv1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nodeName,
+			Namespace: corev1.NamespaceNodeLease,
+		},
+		Spec: coordv1beta1.LeaseSpec{
+			HolderIdentity:       &nodeName,
```

Comment:
This looks wrong.

`nodeName` is the target of the lease, this controller is the holder.
you should be putting something unique to this controller (probably "node-maintenance") in this field.

I think I also mentioned adding to the Node an annotation pointing to this lease object?
This is a pattern used elsewhere for establishing relationships between Nodes/Machines/Hosts.

Reply:
fixed.

Reply:
right now I am not checkingif the lease period has been modified vs that of the lease . Will have to add this check on every reconcile call.

Reply:
added that too.